[SPARK-14774][SQL] Write unscaled values in ColumnVector.putDecimal

sameeragarwal opened the pull request:
## What changes were proposed in this pull request?

We recently made `ColumnarBatch.row` mutable and added a new `ColumnVector.putDecimal` method to support putting `Decimal` values in the `ColumnarBatch`. This unfortunately introduced a bug wherein we were not updating the vector with the proper unscaled values.
## How was this patch tested?

This codepath is hit only when the vectorized aggregate hashmap is enabled. https://github.com/apache/spark/pull/12440 makes sure that a number of regression tests/benchmarks test this bugfix.

sameeragarwal committed:
Fix bug in ColumnVector.putDecimal

sameeragarwal commented:
cc @davies

SparkQA commented:
**[Test build #56425 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56425/consoleFull)** for PR 12541 at commit [`85fdb84`](https://github.com/apache/spark/commit/85fdb84b3f08638ee2714aa65a33cf8552fab3ee).

rxin commented:
We should add a unit test in ColumnVector's own suite.

SparkQA commented:
**[Test build #56425 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56425/consoleFull)** for PR 12541 at commit [`85fdb84`](https://github.com/apache/spark/commit/85fdb84b3f08638ee2714aa65a33cf8552fab3ee).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56425/
Test PASSed.

sameeragarwal committed:
unit test + another bugfix

sameeragarwal commented:
Updated the unit test (we already had one but weren't testing all branches for `DecimalType`). The updated test found another unrelated bug which I also fixed as part of this PR.

SparkQA commented:
**[Test build #56594 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56594/consoleFull)** for PR 12541 at commit [`d2434f5`](https://github.com/apache/spark/commit/d2434f5e69dd0a1d9facf0832755e3ac311aebf9).

davies commented:
LGTM

SparkQA commented:
**[Test build #56594 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56594/consoleFull)** for PR 12541 at commit [`d2434f5`](https://github.com/apache/spark/commit/d2434f5e69dd0a1d9facf0832755e3ac311aebf9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56594/
Test PASSed.

rxin commented:
Merging in master.
